h2o-automl/src/main/java/ai/h2o/automl/TrainingStepsRegistry.java

```diff
@@ -0,0 +1,106 @@
+package ai.h2o.automl;
+
+import ai.h2o.automl.EventLogEntry.Stage;
+import ai.h2o.automl.StepDefinition.Alias;
+import ai.h2o.automl.StepDefinition.Step;
+import water.Iced;
+import water.nbhm.NonBlockingHashMap;
+import water.util.ArrayUtils;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The registry responsible for loading all {@link TrainingStepsProvider} using service discovery,
+ * and providing the list of {@link TrainingStep} to execute.
+ */
+public class TrainingStepsRegistry extends Iced<TrainingStepsRegistry> {
```

Comment:
this is great, it allows for writing user extensions without modifying automl code

- first application could be xgboost, xgboost is built as an h2o-extension and it would be great if automl didn't have to depend on the xgboost module directly

Reply:
@michalkurka I didn't have `xgboost` in mind when doing it, but yes, agree that the XGB steps could be moved to the xgb extension.
Won't do it in this PR, but will try afterwards.